### PR TITLE
r/aws_autoscaling_group: Add support for launch template

### PR DESCRIPTION
**Pull request by gordonbondon**
Adds support to launch ASG from [template](https://docs.aws.amazon.com/autoscaling/ec2/userguide/create-asg-launch-template.html)

I'll file separate PRs for #4267 and #4264 . They are using the same structure.

Tests:
```
→ vaulted -n my -- make testacc TEST=./aws TESTARGS='-run=TestAccAWSAutoScalingGroup_launchTemplate'
Password:
Enter your MFA code: 272073
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAutoScalingGroup_launchTemplate -timeout 120m
=== RUN   TestAccAWSAutoScalingGroup_launchTemplate
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate (103.86s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	104.296s

→ vaulted -n my -- make testacc TEST=./aws TESTARGS='-run=TestAccAWSAutoScalingGroup_WithLoadBalancer'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAutoScalingGroup_WithLoadBalancer -timeout 120m
=== RUN   TestAccAWSAutoScalingGroup_WithLoadBalancer
--- PASS: TestAccAWSAutoScalingGroup_WithLoadBalancer (535.68s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	536.105s

→ vaulted -n my -- make testacc TEST=./aws TESTARGS='-run=TestAccAWSAutoScalingGroup_launchTemplate_toLaunchConfig'
Password:
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAutoScalingGroup_launchTemplate_toLaunchConfig -timeout 120m
=== RUN   TestAccAWSAutoScalingGroup_launchTemplate_toLaunchConfig
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate_toLaunchConfig (144.08s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	144.516s
```

### CODE REVIEW

File: aws/resource_aws_autoscaling_group_test.go [hunk 1]
```diff
@@ -1290,10 +1341,20 @@ resource "aws_elb" "bar" {
...
+    name   = "name"
+    values = ["bitnami-nginxstack-*-linux-debian-9-x86_64-hvm-ebs"]
+  }
+}
+
 resource "aws_launch_configuration" "foobar" {
-  // need an AMI that listens on :80 at boot, this is:
-  // bitnami-nginxstack-1.6.1-0-linux-ubuntu-14.04.1-x86_64-hvm-ebs-ami-99f5b1a9-3
-  image_id = "ami-b5b3fc85"
+  image_id = "${data.aws_ami.test_ami.id}"
```
Comment (bflad) [inline aws/resource_aws_autoscaling_group_test.go]: 😍

**Comment by gordonbondon** [comment]
@bflad almost done. I found a couple of edge cases and tried to resolve them.

After my last changes acceptance test is failing like this:
```
→ vaulted -n my -- make testacc TEST=./aws TESTARGS='-run=TestAccAWSAutoScalingGroup_launchTemplate_update'
Password:
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAutoScalingGroup_launchTemplate_update -timeout 120m
=== RUN   TestAccAWSAutoScalingGroup_launchTemplate_update
--- FAIL: TestAccAWSAutoScalingGroup_launchTemplate_update (60.55s)
	testing.go:518: Step 2 error: Check failed: Check 2/3 error: aws_autoscaling_group.bar: Attribute 'launch_configuration' found when not expected
FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	61.000s
make: *** [testacc] Error 1
```

But when I run the same steps manually I see `"launch_configuration": ""` in `terraform.tfstate`. Maybe you can help me to sort this out.

My other concern is that I don't really like how  `expandLaunchTemplateSpecification` is done, but I don't have an idea on how to do it better. Basic flow should be like this:
- Only one of name or id can be set for LaunchTemplateSpecification but DescribeAustoScaling group returns both of them.
- We need to check that only one of them is set
- When resource is updated check which one has changed and send it
- When none of them has changed - send first available

**Comment by gordonbondon** [comment]
And maybe expandLaunchTemplateSpecification should return interface instead of `*autoscaling.LaunchTemplateSpecification`  so I can somehow make it work for https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#FleetLaunchTemplateSpecification which has the exact structure but another type.

**Comment by bflad** [comment]
I wouldn't try to get fancy merging types. We prefer (mostly) duplicated code over magic/indirection as it makes community maintenance easier.

**Comment by bflad** [comment]
> Only one of name or id can be set for LaunchTemplateSpecification but DescribeAustoScaling group returns both of them.

`Optional: true` and `Computed: true` should handle this situation fine.

> We need to check that only one of them is set

Why? Does the API return an error when both are specified or is this an issue during updates? If it does not error when both are specified, I would assume we're fine passing whichever is defined. If it does error on update because of ID/name mismatch, then we might need to introduce a `CustomizeDiff` function that calls `d.SetNewComputed()` on the opposite attribute

**Comment by gordonbondon** [comment]
On describe it returns:
```
"LaunchTemplate": {
                "LaunchTemplateId": "lt-057f89ebf426f3765",
                "LaunchTemplateName": "test_template",
                "Version": "$Default"
            }
```

**Comment by gordonbondon** [comment]
On describe it returns:
```
"LaunchTemplate": {
                "LaunchTemplateId": "lt-057f89ebf426f3765",
                "LaunchTemplateName": "test_template",
                "Version": "$Default"
            }
```

But if you set both on Create/Update request it will return error:
```
An error occurred (ValidationError) when calling the UpdateAutoScalingGroup operation: You must use a valid fully-formed launch template. A launch template ID and a launch template name cannot be specified in the same request.
```

**Comment by gordonbondon** [comment]
CustomizedDiff and SetnewComputed is great! I was looking for something like this but did not succeeded. I'll try to switch to this approach

**Comment by gordonbondon** [comment]
I've also tried to add plan time validation for attributes like this:
```
func(diff *schema.ResourceDiff, v interface{}) error {
                // Plan time validation for asg instance configuration
                if diff.Id() == "" || !diff.HasChange("launch_configuration") && !diff.HasChange("launch_template") {
                    return nil
                }
                launchConfigurationValue, launchConfigurationOk := diff.GetOk("launch_configuration")
                launchTemplateValue, launchTemplateOk := diff.GetOk("launch_template")
                lc := launchConfigurationOk && launchConfigurationValue.(string) != ""
                lt := launchTemplateOk && len(launchTemplateValue.([]interface{})) > 0
                if lc || lt {
                    return nil
                }
                // return nil
                return errors.New("One of `launch_configuration` or `launch_template` must be set for an autoscaling group")
            },
```

But it does not seem to work when attributes are passed from resources that does not exist yet. Dor axample ``launch_configuration = "${aws_launch_configuration.test.name}"

**Comment by bflad** [comment]
This has been released in [version 1.17.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1170-may-02-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
